symforce/slam/imu_preintegration/manifold_symbolic.py

```diff
@@ -0,0 +1,346 @@
+            in the body frame at timestep i: R_i^T (p_j - p_i - v_i Delta t - 1/2 g Delta t^2),
+            where R_i = pose_i.R, p_i = pose_i.t, and p_j = pose_j.t
+        sqrt_info (sf.M99): sqrt info matrix of DR('s tangent space), Dv, Dp
+        DR_D_gyro_bias (sf.M33): Derivative of DR w.r.t. gyro_bias evaluated at gyro_bias_hat
+        Dp_D_accel_bias (sf.M33): Derivative of Dp w.r.t. accel_bias evaluated at accel_bias_hat
+        Dp_D_gyro_bias (sf.M33): Derivative of Dp w.r.t. gyro_bias evaluated at gyro_bias_hat
+        Dv_D_accel_bias (sf.M33): Derivative of Dv w.r.t. accel_bias evaluated at accel_bias_hat
+        Dv_D_gyro_bias (sf.M33): Derivative of Dv w.r.t. gyro_bias evaluated at gyro_bias_hat
+        accel_bias_hat (sf.V3): The bias of the accelerometer measurements used during pre-integration
+        gyro_bias_hat (sf.V3): The bias of the gyroscope measurements used during pre-integration
+        gravity (sf.V3): Acceleration due to gravity (in the same frame as pose_x and vel_x)
+        dt (T.Scalar): The time between timestep i and j: t_j - t_i
+        epsilon (T.Scalar): epsilon used for numerical stability
+    """
+    delta_gyro_bias = gyro_bias_i - gyro_bias_hat
+    delta_accel_bias = accel_bias_i - accel_bias_hat
+
+    # Correct preintegrated measurements for updated bias estimates
+    corrected_DR = DR * sf.Rot3.from_tangent(
+        (DR_D_gyro_bias * delta_gyro_bias).to_storage(), epsilon
+    )
+    corrected_Dv = Dv + Dv_D_gyro_bias * delta_gyro_bias + Dv_D_accel_bias * delta_accel_bias
+    corrected_Dp = Dp + Dp_D_gyro_bias * delta_gyro_bias + Dp_D_accel_bias * delta_accel_bias
+
+    res_R = (corrected_DR.inverse() * pose_i.R.inverse() * pose_j.R).to_tangent(epsilon)
+    res_v = pose_i.R.inverse() * (vel_j - vel_i - gravity * dt) - corrected_Dv
+    res_p = (
+        pose_i.R.inverse() * (pose_j.t - pose_i.t - vel_i * dt - sf.S(1) / 2 * gravity * dt ** 2)
+        - corrected_Dp
+    )
+
+    res = sf.V9(*res_R, *res_v, *res_p)
+
+    def upper_triangular_mul(ut: sf.M99, v: sf.V9) -> sf.V9:
+        """
+        Returns ut * v. Use to avoid expressions containing the zero-valued lower triangular
+        portion of ut.
+
+        Precondition: ut is upper triangular
+        """
+        return sf.V9([(ut[k, k:] * v[k:, 0])[0] for k in range(9)])
+
+    return upper_triangular_mul(sqrt_info, res)
+
+
+def right_jacobian(tangent: sf.V3, epsilon: T.Scalar) -> sf.M33:
+    """
+    The right jacobian J(v) is d/du Log(Exp(v)^{-1} Exp(v + u)), i.e., a matrix such that
+    Exp(v + dv) ~= Exp(v) Exp(J(v) dv), where v and u are tangent vectors of SO(3).
+
+    Returns J(tangent).
+    """
+    norm = tangent.norm(epsilon)
+    tangent_hat = sf.Matrix.skew_symmetric(tangent)
+    out = sf.M33.eye()
+    out += ((1 - sf.cos(norm)) / (norm ** 2)) * tangent_hat
+    out += ((norm - sf.sin(norm)) / (norm ** 3)) * (tangent_hat * tangent_hat)
+    return out
+
+
+def handwritten_state_D_old_state_gyro_accel(
```

Comment:
In the function name you use state and old_state but in the code, you use new_state and state. Suggest changing the function name to `new_state_D_state_gyro_accel` to be consistent. Or the other way around.

Reply:
Good idea. Changing this to be `new_state_D_state_gyro_accel`.